profiling/src/profile/internal/string_table.rs

```diff
@@ -0,0 +1,227 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2023-Present Datadog, Inc.
+
+use super::*;
+use bumpalo::Bump;
+use ouroboros::self_referencing;
+use std::collections::HashMap;
+use std::hash::BuildHasherDefault;
+
+#[cfg(test)]
+use std::ops::Range;
+
+struct BorrowedStringTable<'b> {
+    /// The arena to store the characters in.
+    arena: &'b Bump,
+
+    /// Used to have efficient lookup by [StringId], and to provide an
+    /// [Iterator] over the strings.
+    vec: Vec<&'b str>,
+
+    /// Used to have efficient lookup by [&str].
+    map: HashMap<&'b str, StringId, BuildHasherDefault<rustc_hash::FxHasher>>,
+}
+
+#[self_referencing]
+struct StringTableCell {
+    // This arena holds the characters of the strings. The memory used to hold
+    // the vec and map to implement the lookups required are not in the arena,
+    // because they would be re-allocated over time but the previous data
+    // wouldn't generally get reclaimed. This makes them a poor fit for an
+    // arena allocator.
+    owner: Bump,
+
+    // This says that the BorrowedStringTable will hold a reference to the
+    // field `owner`. Rust does not allow this as there are many ways to make
+    // this code unsafe. The ouroboros crate is used to provide a safe
+    // abstraction for a subset of self-referential behavior.
+    #[borrows(owner)]
+    #[covariant]
+    dependent: BorrowedStringTable<'this>,
+}
+
+/// The [StringTable] stores strings and associates them with [StringId]s,
+/// which correspond to the order in which strings were inserted. The empty
+/// string is always associated with [StringId::ZERO].
+pub struct StringTable {
+    // ouroboros will add a lot of functions to this struct, which we don't
+    // want to expose publicly, so the internals are wrapped and private.
+    inner: StringTableCell,
+}
+
+impl StringTable {
+    #[inline]
+    pub fn new() -> Self {
+        Self::with_capacity(4000)
```

Comment:
why 4000?

Reply:
The goal is to get the system allocator to use large, good allocations. I could see this being `4096`, but due to the overhead, it will actually allocate something larger like 8192 - 64. Not sure if that's better, could be.

Reply:
At least on some glibc versions (search `man mallopt`), the point malloc starts using `mmap` is 128 * 1024 bytes (128KiB). So if we knew some statistics about how large string tables are in real-world profiles, grouped by language, we could possibly bump this to be near this limit and see improved performance, but also the memory will definitely get returned to the OS.

I guess I could start by analyzing a symfony-demo profile, this gives us a floor (not special, on the small side for PHP profiles almost assuredly).

Reply:
A symfony-demo profile I analyzed came out with 1611 strings, which used at least 101,342 bytes of memory (for the string data, not any other kind of overhead in the vecs themselves or anything).

Reply:
I documented many things and bumped this number to 8 pages, or 32,768 bytes. Feel free to argue for something different, but without metrics from intake or something, it's hard to say what it ought to be.